opm/simulators/utils/GridDataOutput.hpp

```diff
@@ -0,0 +1,455 @@
+     *       should contain at least as many vlaues as the mesh has cells  (ncells_) or vertices (nvertices_)
+     *       depending on if data is cell centred or vertex centred, respectively.
+     *
+     *  As we are templated on the Dune::PartitionSet<partitions>, values for ncorners_, nvertices_ and ncells_ cannot change
+     *
+     *  This class does not work with grids containing polyhedral cells (well, it has not been tested
+     *  with this kind of grid data). The user should call polyhedralCellPresent() to test if polyhedral 
+     *  cells are present and decide what they want to do before copying data using the data accessor methods.
+     */
+    explicit SimMeshDataAccessor ( const GridView &gridView,  
+                            Dune::PartitionSet<partitions>  dunePartition)
+        : gridView_( gridView ),
+          dunePartition_(dunePartition)
+    {
+        dimw_ = GridView::dimension ; // this is an enum
+        partition_value_ = dunePartition.value ;
+        countEntities() ;
+    }
+
+    //! destructor
+    ~SimMeshDataAccessor ()
+    {
+    }
+
+     /**
+       Checks for cells that have polyhedral type within the current partition of cells
+       
+       Returns true if a polyhedral sell is found. If this is the case then this partition 
+       is not going to be available for visualisation as this class does not yet handle 
+       polyhedral cells.
+    */
+    bool polyhedralCellPresent() 
+    {
+        for (const auto& cit :  elements(gridView_, dunePartition_))
+        {
+            auto  corner_geom = cit.geometry() ;
+            if( Dune::VTK::geometryType( corner_geom.type() ) == Dune::VTK::polyhedron )
+            {
+                return true ;
+            }
+       } 
+       return false;
+    }
+
+    /** 
+        Count the vertices, cells and corners.
+
+        Count all the vertices ( the Dune::Partitions::all partition ) as then we do not need to renumber
+        the vertices as all the subsets use references to the full set.
+    */ 
+    void countEntities(  )
+    {
+        // We include all the vertices for this ranks partition
+        const auto& vert_partition_it = vertices(gridView_, Dune::Partitions::all);
+        nvertices_ = std::distance(vert_partition_it.begin(), vert_partition_it.end());
+
+        const auto& cell_partition_it = elements(gridView_, dunePartition_);
+        ncells_ = std::distance(cell_partition_it.begin(), cell_partition_it.end());
+
+        ncorners_ = 0 ;
+        for (const auto& cit : cell_partition_it)
+        {
+            auto  corner_geom = cit.geometry() ;
+            ncorners_ += corner_geom.corners() ;
+        }
+    }
+
+     /**
+       Write the positions of vertices - directly to the pointers given in paramaters 
+
+       Returns the number of vertices written
+    */
+    template <typename T> 
+    long  writeGridPoints( T*  x_inout,  T*  y_inout, T* z_inout )
```

Comment:
It may be a more simple option to add an expected/maximum size available argument to the method:
```
long  writeGridPoints( T*  x_inout,  T*  y_inout, T* z_inout, long max_size )
```

I currently have issues with the type returned from the decltype() : 

using VT = decltype(x_inout.data()[0]);

 error: invalid ‘static_cast’ from type ‘double’ to type ‘VT’ {aka ‘double&’}
  206 |                 z_inout.data()[i] = static_cast<VT>(0.0);
      |                                     ^~~~~~~~~~~~~~~~~~~~

Any idea why it ends up as a reference and not just a POD type?

Reply:
okay, fixed the &double issue by using:
```
double td = 0.0 ;
z_inout.data()[i] = static_cast<VT>(td);
```

Reply:
> ```
> using VT = decltype(x_inout.data()[0]);
> 
>  error: invalid ‘static_cast’ from type ‘double’ to type ‘VT’ {aka ‘double&’}
>   206 |                 z_inout.data()[i] = static_cast<VT>(0.0);
>       |                                     ^~~~~~~~~~~~~~~~~~~~
> ```
I assume there's a good reason for going through `.data()` instead of just subscripting the object directly.  Otherwise, it'd just be easier to write
```
z_inout[i] = static_cast<VT>(0);
```

As for why you're getting diagnostic messages about conversions from `double&` to `double`, that has to do with the exact expression you're using to define `VT`.  Since `x_inout` is a `T*`, meaning you can both read from it and write to it unless `T` itself happens to be `const`, the result of the subscript operator will always be a reference.  The common way of getting the "pure" underlying type is
```C++
using VT = std::remove_cv_t<std::remove_reference_t<decltype(...)>>;
```
with [`std::remove_reference<>`](https://en.cppreference.com/w/cpp/types/remove_reference) and [`std::remove_cv<>`](https://en.cppreference.com/w/cpp/types/remove_cv) both being defined in the `<type_traits>` header.  In C++20 we can use the "fused" type-trait [`std::remove_cvref_t<>`](https://en.cppreference.com/w/cpp/types/remove_cvref) as a short-hand for the above.

That said, I agree with @atgeirr that taking `T&` with the type `T` having a subcript operator and a `size()` member function would be the idiomatic C++ interface.

Reply:
Sorry for forgetting that the T would be a reference in my example, you indeed need to strip that from the type.

Reply:
I have got methods accepting templated Vector clones (a DamarisVar type with size() and data() methods). The code is compiling and running. The DamarisVar vector clone type is fraught with danger as the actual size cannot be queried reliably, and the size is governed by the parameters used in the XML file. It works, but it is just as unsafe as the original code, or a version that adds a max_size parameter to the data filling method. We have pretty good control of  the XML file, as we build it programmatically, so it is not such a problem.

Reply:
I have updated the code docstrings and have given an example of using the vector version of the methods.